### PR TITLE
RetroPlayer: Include video frame in savestate

### PR DESCRIPTION
**Pull request by garbear**
## Description

This change is the final one broken out from https://github.com/xbmc/xbmc/pull/22635.

The change here extends our savestate format to include pixels. This allows us to render savestates instantly using their original pixels, instead of a degraded jpg slowly written to disk and loaded back, missing all frame metadata such as rotation and aspect ratio.

It also improves the responsiveness of creating savestates in-game. Instead of a ~quarter second delay, savestates are now created and shown in the GUI instantly.

The savestate version is bumped from 2 to 3. The version change is fully backwards- and forwards-compatible. In older versions, the pixel data is safely ignored, and in newer versions, Kodi falls back to the jpg screenshot.

Even if Savestate Rendering is deferred to a later version, it's a good idea to get pixel data in savestates as soon as possible so that it'll have something to render when its ready.

In brief testing, the size of SNES savestates increased from ~810 KB to ~1 MB. In my development branch I've already fully added lz4 as a dependency, so the size increase will be trivial once we implement simple savestate compression.

The only thing I'm not happy about, is that I'd like to move the video properties out of savestate.fbs and into video.fbs, but I'm planning a backwards-incompatible change to savestates next (though in the distant future), so I'm not sure if it's worth moving the fields now. If I change my mind I'll update the PR before merge.

## Motivation and context

Get savestates ready for savestate rendering in https://github.com/xbmc/xbmc/pull/22635.

## How has this been tested?

Included in test builds: https://github.com/garbear/xbmc/releases

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp [hunk 1]
```diff
@@ -74,6 +80,13 @@ void CRPRenderManager::Deinitialize()
     buffer->Release();
   m_pendingBuffers.clear();
 
+  for (auto it : m_savestateBuffers)
```
Comment (lrusak) [inline xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp:83]:
```suggestion
  for (auto& it : m_savestateBuffers)
```

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp [hunk 2]
```diff
@@ -74,6 +80,13 @@ void CRPRenderManager::Deinitialize()
...
+  {
+    for (auto renderBuffer : it.second)
```
Comment (lrusak) [inline xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp:85]:
```suggestion
    for (auto& renderBuffer : it.second)
```

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp (hunk 1 above)
Comment (lrusak) [reply xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp:83]: you could also possibly do something cool like:

```suggestion
  for (auto& [saveStatePath, renderBuffers] : m_savestateBuffers)
```

**Review by lrusak — APPROVED** [review]
Looks good. This is really awesome I can't wait to have this in action.

I wish we could get away from using unsized types but it's so engrained in Kodi that I can look past it.

**Comment by garbear** [comment]
I really wanted the video metadata in a separate table defined in `video.fbs`, but I'm not able to get the table-within-table code to not crash, and it's hard to find documentation or examples about table-within-tables, and it's even suggested that Python doesn't support table-within-tables, so I think I abandon this effort.

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp (hunk 2 above)
Comment (garbear) [reply xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp:85]: `renderBuffer` is of type `IRenderBuffer*` so it doesn't need to assign by reference, correct?

**Comment by garbear** [comment]
jenkins build this please

**Comment by garbear** [comment]
jenkins build this please

**Pull request merged**